Sync Merge: update-workshop-to-c5 (#797)
espressif/developer-portal: content/workshops/esp-idf-with-esp32-c5/lecture-3/index.md

Besides the already mentioned (`data`,`nvs`), this partition table contains a (`data`,`ota`) field which plays an important role in OTA updates.

#### otadata partition

The `otadata` partition is a special partition in the ESP-IDF partition table, required for projects that use Over-The-Air (OTA) firmware updates. Its main purpose is to store information about which OTA app slot (such as `ota_0` or `ota_1`) should be booted by the device. It's typical size is `0x2000` bytes (two flash sectors)

The otadata partition is used as follows:
* On first boot (or after erasing), the otadata partition is empty (all bytes set to 0xFF). In this state, the bootloader will boot the factory app if present, or the first OTA slot if not.
* After a successful OTA update, the otadata partition is updated to indicate which OTA app slot should be booted next.
* The partition is designed to be robust against power failures: it uses two sectors, and a counter field to determine the most recent valid data if the sectors disagree.

> ## Conclusion

Reviewer: I think it's missing the OTA rollback prevention feature.

https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/ota.html#app-rollback

Reviewer 2: added section about rollback.

In the OTA process, the Espressif device downloads the firmware from a given location, as depicted in Fig.1.

{{< figure
default=true
src="https://developer.espressif.com/workshops/esp-idf-advanced/assets/lecture-4-ota-diagram.webp"
width=350
caption="Fig.1 -- OTA basic diagram"
    >}}

Key benefits of OTA include:

* __Remote maintenance:__ Update firmware without on-site visits.
* __Improved security:__ Quickly patch known vulnerabilities.
* __Feature updates:__ Seamlessly deliver new functionality to users.
* __Lower maintenance costs:__ Avoid expensive manual recalls or servicing.

### Implementing OTA with ESP-IDF

ESP-IDF offers built-in support for OTA through two main methods:

* __Native API__: Using the `app_update` component for full control over the update process.
* __Simplified API__: Using the `esp_https_ota` component for a higher-level interface that handles HTTPS download and flashing automatically.

> In most cases, application needs to interact with public interface of `esp_https_ota` and `app_update` components only.

Reviewer 3: ```suggestion
In most cases, application needs to interact with the public interfaces of `esp_https_ota` and `app_update` components only.
```

{{< /alert >}}

#### Example code snippet using `esp_https_ota`

Using `esp_https_ota` is straightforward and typically requires just a few lines of code.

```c
#include "esp_https_ota.h"

esp_err_t do_firmware_upgrade()
{
    esp_http_client_config_t config = {
        .url = "https://example.com/firmware.bin",
        .cert_pem = (char *)server_cert_pem_start,
    };
    esp_https_ota_config_t ota_config = {
        .http_config = &config,
    };
    esp_err_t ret = esp_https_ota(&ota_config);
    if (ret == ESP_OK) {
        esp_restart();
    } else {
        return ESP_FAIL;
    }
    return ESP_OK;
}
```
This code downloads a new firmware image and, if successful, restarts the device to boot into the new firmware. For more advanced usage, refer to the [ESP-IDF OTA documentation](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/ota.html).

#### OTA partition table layout

OTA requires a specific partition table layout. At minimum, you need:

* __NVS partition:__ For non-volatile storage.
* __`otadata` partition:__ To track which firmware partition is active.
* __Two OTA app partitions:__ For active/passive firmware images.

> An example of valid partition table is the following.

Reviewer 3: ```suggestion
An example of a valid partition table is as follows:
```

```
Name,   Type, SubType, Offset,  Size, Flags
nvs,      data, nvs,     ,        0x6000,
otadata,  data, ota,     ,        0x2000,
phy_init, data, phy,     ,        0x1000,
ota_0,    app,  ota_0,   ,        1M,
ota_1,    app,  ota_1,   ,        1M,
```

> This layout ensures safe updates: the new firmware is written to the inactive partition, and only after verification is it marked as active for the next boot. The OTA data partition is two flash sectors (0x2000 bytes) to prevent corruption in case of power failure during updates.

Reviewer 3: ```suggestion
This layout ensures safe updates: the new firmware is written to the inactive partition, and only after verification is it marked as active for the next boot. The OTA data partition is two flash sectors (`0x2000` bytes) to prevent corruption in case of power failure during updates.
```

Besides the already mentioned (`data`,`nvs`), this partition table contains a (`data`,`ota`) field which plays an important role in OTA updates.

#### otadata partition

The `otadata` partition is a special partition in the ESP-IDF partition table, required for projects that use Over-The-Air (OTA) firmware updates. Its main purpose is to store information about which OTA app slot (such as `ota_0` or `ota_1`) should be booted by the device. It's typical size is `0x2000` bytes (two flash sectors)

The otadata partition is used as follows:

> * On first boot (or after erasing), the otadata partition is empty (all bytes set to 0xFF). In this state, the bootloader will boot the factory app if present, or the first OTA slot if not.

Reviewer 3: ```suggestion
* On first boot (or after erasing), the otadata partition is empty (all bytes set to `0xFF`). In this state, the bootloader will boot the factory app if present. If not, it will boot the first OTA slot.
```

> * After a successful OTA update, the otadata partition is updated to indicate which OTA app slot should be booted next.

Reviewer 3: ```suggestion
* After a successful OTA update, the `otadata` partition is updated to indicate which OTA app slot should be booted next.
```
